pkg/server/api/localauthority/v1/service.go

```diff
@@ -0,0 +1,279 @@
+	return nil, status.Error(codes.Unimplemented, "unimplemented")
+}
+
+func (s *Service) RevokeJWTAuthority(context.Context, *localauthorityv1.RevokeJWTAuthorityRequest) (*localauthorityv1.RevokeJWTAuthorityResponse, error) {
+	return nil, status.Error(codes.Unimplemented, "unimplemented")
+}
+
+func (s *Service) GetX509AuthorityState(ctx context.Context, _ *localauthorityv1.GetX509AuthorityStateRequest) (*localauthorityv1.GetX509AuthorityStateResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	var states []*localauthorityv1.AuthorityState
+	current := s.ca.GetCurrentX509CASlot()
+	if !current.IsEmpty() {
+		if current.AuthorityID() == "" {
+			return nil, api.MakeErr(log, codes.Internal, "current slot does not contains authority ID", nil)
+		}
+
+		states = append(states, &localauthorityv1.AuthorityState{
+			AuthorityId: current.AuthorityID(),
+			Status:      localauthorityv1.AuthorityState_ACTIVE,
+		})
+	}
+
+	next := s.ca.GetNextX509CASlot()
+	// when next has a key indicates that it was initialized
+	if next.AuthorityID() != "" {
+		status := localauthorityv1.AuthorityState_PREPARED
+		// CA is removed from slot on rotation
+		if next.IsEmpty() {
+			status = localauthorityv1.AuthorityState_OLD
+		}
+
+		states = append(states, &localauthorityv1.AuthorityState{
+			AuthorityId: next.AuthorityID(),
+			Status:      status,
+		})
+	}
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.GetX509AuthorityStateResponse{
+		States: states,
+	}, nil
+}
+
+func (s *Service) PrepareX509Authority(ctx context.Context, req *localauthorityv1.PrepareX509AuthorityRequest) (*localauthorityv1.PrepareX509AuthorityResponse, error) {
+	log := rpccontext.Logger(ctx)
+
+	if err := s.ca.PrepareX509CA(ctx); err != nil {
+		return nil, api.MakeErr(log, codes.Internal, "failed to prepare X.509 authority", err)
+	}
+
+	slot := s.ca.GetNextX509CASlot()
+	// Prepare is going to use current slot when it is empty
+	if slot.IsEmpty() {
+		slot = s.ca.GetCurrentX509CASlot()
+	}
+
+	rpccontext.AuditRPC(ctx)
+
+	return &localauthorityv1.PrepareX509AuthorityResponse{
+		PreparedAuthority: &localauthorityv1.AuthorityState{
+			AuthorityId: slot.AuthorityID(),
+			Status:      localauthorityv1.AuthorityState_PREPARED,
+		},
+	}, nil
+}
+
+func (s *Service) ActivateX509Authority(ctx context.Context, req *localauthorityv1.ActivateX509AuthorityRequest) (*localauthorityv1.ActivateX509AuthorityResponse, error) {
+	rpccontext.AddRPCAuditFields(ctx, buildAuditLogFields(req.AuthorityId))
+	log := rpccontext.Logger(ctx)
+
+	// TODO: implement a way to activate an OLD authority
+	if req.AuthorityId != "" {
+		log = log.WithField(telemetry.LocalAuthorityID, req.AuthorityId)
+		return nil, api.MakeErr(log, codes.InvalidArgument, "activating an old authority is not supported yet", nil)
+	}
+
+	if s.ca.GetNextX509CASlot().IsEmpty() {
+		return nil, api.MakeErr(log, codes.Internal, "no prepared authority found", nil)
+	}
```

Comment:
Good check here ... I was surprised to see that CA manager doesn't do this check .. maybe it should?

Reply:
not sure if this should validate this on Activate, since we depends on dates, on normal rotations this must never happens, the issue here is that basically we are forcing this, and preparation may not be done yet,
I prefer to return error here since we allow users to force a preparation and be sure that the prepared bundle is propagated before they activate it.